documentation/general/dotnet-run-file.md

```diff
@@ -0,0 +1,266 @@
+> and also it might not be clear to users that `dotnet run file.cs` will include other `.cs` files in the compilation.
+> Therefore we could consider some switch (a command-line option and/or a `#` language directive) to enable/disable this behavior.
+> When disabled, [grow up](#grow-up) would generate projects in subdirectories similarly to [multi-entry-point scenarios](#multiple-entry-points)
+> to preserve the behavior.
+
+### Nested files
+
+If there are nested project files like
+```
+App/File.cs
+App/Nested/Nested.csproj
+App/Nested/File.cs
+```
+executing `dotnet run app/file.cs` includes the nested `.cs` file in the compilation.
+That might be unexpected, hence we could consider reporting an error in such situation.
+However, the same problem exists for normal builds with explicit project files
+and usually the build fails because there are multiple entry points or other clashes.
+
+Similarly, we could report an error if there are many nested directories and files,
+so for example if someone puts a C# file into `C:/`
+and executes `dotnet run C:/file.cs` or opens that in the IDE, we do not walk the whole drive.
+Again, this problem exists with project-based programs as well, albeit it's likely uncommon there
+since creating projects with `dotnet new` or an IDE creates them in a directory.
+
+### Multiple entry points
+
+If there are multiple entry-point files in the target directory, the target path must be a file
+(an error is reported if it points to a directory instead).
+Then the build ignores other entry-point files.
+
+Thanks to this, it is possible to have a structure like
+```
+App/Util.cs
+App/Program1.cs
+App/Program2.cs
+```
+where either `Program1.cs` or `Program2.cs` can be run and both of them have access to `Util.cs`.
+
+In this case, there are multiple implicit projects
+(and during [grow up](#grow-up), multiple project files are materialized
+and the original C# files are moved to the corresponding project subdirectories):
+```
+App/Shared/Util.cs
+App/Shared/Shared.csproj
+App/Program1/Program1.cs
+App/Program1/Program1.csproj
+App/Program2/Program2.cs
+App/Program2/Program2.csproj
+```
+
+The generated folders might need to be named differently to avoid clashes with existing folders.
+
+The entry-point projects (`Program1` and `Program2` in our example)
+have a `ProjectReference` to the non-entry-point `Shared` project.
+
+Unless the [artifacts output layout][artifacts-output] is used (which is recommended),
+those implicit projects mean that build artifacts are placed under those implicit directories
+even though they don't exist on disk prior to build:
+```
+App/Shared/bin/
+App/Shared/obj/
+App/Program1/bin/
+App/Program1/obj/
+App/Program2/bin/
+App/Program2/obj/
+```
+
+## Package references
+
+It is possible to specify NuGet package references via the `#r` directive.
+
+```cs
+#r "nuget: Newtonsoft.Json, 13.0.1"
```

Comment:
We should update this to the latest proposal of `#package Package.Id 0.1.1` yes?